src/areastore.cpp

```diff
@@ -0,0 +1,328 @@
+	}
+	return false;
+}
+
+void VectorAreaStore::getAreasForPosImpl(std::vector<Area *> *result, v3s16 pos)
+{
+	size_t msiz = m_areas.size();
+	for (size_t i = 0; i < msiz; i++) {
+		Area *b = m_areas[i];
+		if (AST_CONTAINS_PT(b, pos)) {
+			result->push_back(b);
+		}
+	}
+}
+
+void VectorAreaStore::getAreasInArea(std::vector<Area *> *result,
+		v3s16 minedge, v3s16 maxedge, bool accept_overlap)
+{
+	size_t msiz = m_areas.size();
+	for (size_t i = 0; i < msiz; i++) {
+		Area * b = m_areas[i];
+		if (accept_overlap ? AST_AREAS_OVERLAP(minedge, maxedge, b) :
+				AST_CONTAINS_AREA(minedge, maxedge, b)) {
+			result->push_back(b);
+		}
+	}
+}
+
+bool VectorAreaStore::forEach(bool (*callback)(void *args, Area *a), void *args) const
+{
+	size_t msiz = m_areas.size();
+	for (size_t i = 0; i < msiz; i++) {
+		if (callback(args, m_areas[i])) {
+			return true;
+		}
+	}
+	return false;
+}
+
+#if USE_SPATIAL
+
+static inline SpatialIndex::Region get_spatial_region(const v3s16 minedge,
+		const v3s16 maxedge)
+{
+	const double p_low[] = {(double)minedge.X,
+		(double)minedge.Y, (double)minedge.Z};
+	const double p_high[] = {(double)maxedge.X, (double)maxedge.Y,
+		(double)maxedge.Z};
+	return SpatialIndex::Region(p_low, p_high, 3);
+}
+
+static inline SpatialIndex::Point get_spatial_point(const v3s16 pos)
+{
+	const double p[] = {(double)pos.X, (double)pos.Y, (double)pos.Z};
+	return SpatialIndex::Point(p, 3);
+}
+
+
+void SpatialAreaStore::insertArea(const Area &a)
+{
+	areas_map[a.id] = a;
+	m_tree->insertData(0, NULL, get_spatial_region(a.minedge, a.maxedge), a.id);
+	invalidateCache();
+}
+
+bool SpatialAreaStore::removeArea(u32 id)
+{
+	std::map<u32, Area>::iterator itr = areas_map.find(id);
+	if (itr != areas_map.end()) {
+		Area *a = &itr->second;
+		bool result = m_tree->deleteData(get_spatial_region(a->minedge,
+			a->maxedge), id);
+		invalidateCache();
+		return result;
+	} else {
+		return false;
+	}
+}
+
+void SpatialAreaStore::getAreasForPosImpl(std::vector<Area *> *result, v3s16 pos)
+{
+	VectorResultVisitor visitor(result, this);
+	m_tree->pointLocationQuery(get_spatial_point(pos), visitor);
+}
+
+void SpatialAreaStore::getAreasInArea(std::vector<Area *> *result,
+		v3s16 minedge, v3s16 maxedge, bool accept_overlap)
+{
+	VectorResultVisitor visitor(result, this);
+	if (accept_overlap) {
+		m_tree->intersectsWithQuery(get_spatial_region(minedge, maxedge),
+			visitor);
+	} else {
+		m_tree->containsWhatQuery(get_spatial_region(minedge, maxedge), visitor);
+	}
+}
+
+bool SpatialAreaStore::forEach(bool (*callback)(void *args, Area *a), void *args) const
+{
+	// TODO ?? (this is only needed for serialisation, but libspatial has its own serialisation)
+	return false;
+}
+
+SpatialAreaStore::~SpatialAreaStore()
+{
+	delete m_tree;
+}
+
+SpatialAreaStore::SpatialAreaStore()
+{
+	m_storagemanager =
+		SpatialIndex::StorageManager::createNewMemoryStorageManager();
+	SpatialIndex::id_type id;
+	m_tree = SpatialIndex::RTree::createNewRTree(
```

Comment:
One line for each parameter??
Maybe this would be a good idea if you annotated what each parameter was in a comment, but this merely takes up a lot of space as-is.